A55-xds-stateful-session-affinity.md

```diff
@@ -0,0 +1,480 @@
+`xds_cluster_resolver_experimental` config.
+
+Also note that [common_lb_config][] field is incompatible with the new
+[load_balancing_policy][] field used for custom LB policies (see
+[A52: gRPC xDS Custom Load Balancer Configuration][grfc_a52]) which means
+custom LB policies and stateful session affinity features will be incompatible
+with each other. This will need to be fixed eventually to be able to use these
+two features together. Until that happens, we assume the default value
+for the [`common_lb_config.override_host_status`][or-host-status] field
+when one is using the new [load_balancing_policy][] field so the user is still
+able to use the feature.
+
+### `CookieBasedStatefulSessionFilter` Processing
+
+A channel configured for stateful session affinity will have the
+`CookieBasedStatefulSessionFilter` installed (interceptor in Java/Go).
+This filter will process incoming RPCs and set an appropriate LB pick value
+which will be passed to the Load Balancing Policy's Pick method.
+
+When an RPC arrives on a configured channel, the Config Selector processes it
+as described in [A31: gRPC xDS Config Selector Design][A31]. This includes the
+`CookieBasedStatefulSessionFilter` which is installed as one of the filters in
+the channel. Note that the filter maintains a context for an RPC and processes
+both the outgoing request and incoming response of the RPC within that context.
+Also note that the filter is initialized with 3 configuration values - `name`,
+`ttl`  and `path` as described above.
+
+The filter performs the following steps:
+
+* If the incoming RPC “path” does not match the `path` configuration value,
+skip any further processing (including response processing) and just pass the
+RPC through (to the next filter). For example if the `path` configuration
+value is `/Package1.Service2` and the RPC method is
+`/Package2.Service1/Method3` then just pass the RPC through. Note, for path
+matching the rules described in [rfc6265 section-5.1.4][rfc6265-path-match]
+will apply.
+
+* Search the RPC headers (metadata) for header(s) named `Cookie` and get the
+set of all the cookies present. If the set does not contain a cookie whose
+name matches the `name` configuration value of the filter, then pass the RPC
+through. If you find more than one cookie with that `name` then log a local
+warning and pass the RPC through. Otherwise get the value of that cookie and
+base64-decode the value to get the `override-host` for the RPC. This value
+should be a syntactically valid IP:port address: if not, then log a local
+warning and pass the RPC through. For a valid `override-host` value, set it
+in the RPC context as the value of `upstreamAddress` state variable which
+will be used in the response processing of that RPC.
+
+* In case of a valid `override-host` value, pass this value as a
+"pick-argument" to the Load Balancer's pick method as described in [A31][].
+For example,
+
+    * in Java use `CallOptions` to pass this value.
```

Comment:
To confirm: what will happen if the user sets this in Context ? It would be really great to allow them to do it- will cover additional modes of discovery and avoid the need to do the ugly hacks I descirbed ( connect to the IP directly and do all tls/auth manually, bypassing XDS ). Users can get the IP:port from k8s, multicast or remote address of a prior connection.